processor/spanmetricsprocessor/processor.go

```diff
@@ -249,39 +271,56 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
 // writes the raw metrics data into the metrics object.
 func (p *processorImp) buildMetrics() (*pdata.Metrics, error) {
 	m := pdata.NewMetrics()
-	ilm := m.ResourceMetrics().AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
-	ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+	rms := m.ResourceMetrics()
+	for _, key := range p.resourceKeyToDimensions.Keys() {
+		cachedResourceAttributesMap, ok := p.resourceKeyToDimensions.Get(key)
+		if !ok {
+			return nil, errors.New("expected cached resource attributes not found")
+		}
 
-	// Obtain write lock to reset data
-	p.lock.Lock()
+		resourceAttributesMap, ok := cachedResourceAttributesMap.(pdata.AttributeMap)
+		if !ok {
+			return nil, errors.New("expected cached resource attributes type assertion failed")
+		}
 
-	if err := p.collectCallMetrics(ilm); err != nil {
-		p.lock.Unlock()
-		return nil, err
-	}
+		// If the service name doesn't exist, we treat it as invalid and do not generate a trace
+		if _, exist := resourceAttributesMap.Get(serviceNameKey); !exist {
+			continue
+		}
 
-	if err := p.collectLatencyMetrics(ilm); err != nil {
-		p.lock.Unlock()
-		return nil, err
-	}
+		rm := rms.AppendEmpty()
 
-	p.metricKeyToDimensions.RemoveEvictedItems()
+		// Iterate over `AttributeMap` structure defining resource attributes to append to the metric resource and append
+		resourceAttributesMap.Range(func(k string, v pdata.AttributeValue) bool {
+			rm.Resource().Attributes().Insert(k, v)
+			return true
+		})
 
-	// If delta metrics, reset accumulated data
-	if p.config.GetAggregationTemporality() == pdata.MetricAggregationTemporalityDelta {
-		p.resetAccumulatedMetrics()
-	}
-	p.resetExemplarData()
+		ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+		ilm.InstrumentationLibrary().SetName(instrumentationLibraryName)
+
+		// build metrics per resource
+		resourceAttrKey, ok := key.(resourceKey)
+		if !ok {
+			return nil, errors.New("resource key type assertion failed")
+		}
 
-	p.lock.Unlock()
+		if err := p.collectCallMetrics(ilm, resourceAttrKey); err != nil {
+			return nil, err
+		}
+
+		if err := p.collectLatencyMetrics(ilm, resourceAttrKey); err != nil {
+			return nil, err
+		}
 
+	}
 	return &m, nil
 }
 
 // collectLatencyMetrics collects the raw latency metrics, writing the data
 // into the given instrumentation library metrics.
-func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMetrics) error {
-	for key := range p.latencyCount {
+func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMetrics, resAttrKey resourceKey) error {
+	for mKey := range p.latencyCount[resAttrKey] {
```

Comment:
`collectLatencyMetrics()` should avoid being used concurrently as this will cause incorrect aggregation. As it's an internal function that should not be used concurrently, we didn't add protection here. However, we added the lock at the top level to avoid concurrent usages.